### PR TITLE
vdk-meta-jobs: exec job with arguments

### LINKED ISSUE #1502: Executing job from another job (synchronously) with arguments
**What is the feature request? What problem does it solve?**
I want to execute a job from another job synchronously with specifying arguments. Then I need to be able to get the result (i.e. success or failure).

**Suggested solution**
It'd be similar to the meta job plugin - use execute API and track the result.

**Additional context**
I would use this to extract reusable critical section logic.

### PR DESCRIPTION
**Pull request by yonitoo**
What:
Introduce execution of jobs with arguments (synchronously) via DAGs.

The way a job dict looks now:
{
        "job_name": "name-of-job",
        "team_name": "team-of-job",
        "fail_meta_job_on_error": True or False,
        **"arguments": \<json\>**,
        "depends_on": [name-of-job1, name-of-job2]
}

GitHub Issue: [https://github.com/vmware/versatile-data-kit/issues/1502](https://github.com/vmware/versatile-data-kit/issues/1502)

Testing Done: existing tests + new unit test

Closes #1502

### CODE REVIEW

**Comment by tozka** [comment]
Please update your PR with latest changes. There is  a lot of unrelated changes in the PR.

**Comment by yonitoo** [comment]
Done. It is ok now.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py [hunk 1]
```diff
@@ -141,6 +149,28 @@ def _validate_fail_meta_job_on_error(self, job: Dict):
...
+    def _validate_arguments(self, job: Dict):
+        if "arguments" in job:
+            if not isinstance(job["arguments"], Dict):
+                self._raise_error(
+                    list(job["job_name"]),
+                    ERROR.TYPE,
+                    "The type of the job dict key arguments is not str.",
+                    f"Change the Data Job Dict value of arguments. "
+                    f"Current type is {type(job['arguments'])}. Expected type is JSON-formatted str.",
+                )
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:161]: You are checking if job["arguments"} is Dict, but then saying it must be a string. 

It should be a python dictionary by the way. Not a string. It's a not a nice thing to force user to format their arguments as a string. In Python of course we should be using pythonic objects. 

Better validation is to check if it can be serialized as json. 

So as a validation running `json.dumps` and not loads .

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py [hunk 2]
```diff
@@ -22,7 +22,12 @@
 
 
 class MetaJobsDag:
-    def __init__(self, team_name: str, meta_config: MetaPluginConfiguration):
+    def __init__(
+        self,
+        team_name: str,
+        meta_config: MetaPluginConfiguration,
+        arguments: Dict = None,
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py:29]: This looks wrong. 

Arguments must be per job and not for the whole Dag

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py [hunk 3]
```diff
@@ -49,12 +49,14 @@ def __init__(
...
         team_name: str,
         rest_api_url: str,
+        arguments: dict = None,
         timeout: int = 5,  # TODO: Set reasonable default
         **kwargs,
     ) -> None:
         self.job_name = job_name
         self.team_name = team_name
         self._rest_api_url = rest_api_url
+        self.arguments = arguments
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py:59]: should be private . All fields here should be marked as private per https://github.com/vmware/versatile-data-kit/wiki/Coding-Standard#clarification-on-access-modifiers-in-python

File: projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py [hunk 4]
```diff
@@ -327,6 +332,11 @@ def test_meta_job_duplicate_jobs(self):
     def test_meta_job_not_allowed_job_key(self):
         self._test_meta_job_validation("meta-job-not-allowed-job-key")
 
+    def test_meta_job_wrong_job_arguments_json(self):
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py:335]: wher is the happy path tested ? That you successfully passed arguments to the job ?

You should test a few happy paths : 
* Where arguments are missing - None (which is already tested by previous test
* Where arguments is empty {} 
* Norma {"key: value" , "key2": value}

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta.py [hunk 5]
```diff
@@ -11,9 +12,10 @@
 
 class IDataJobExecutor(abc.ABC):
     @abc.abstractmethod
-    def start_job(self, job_name: str, team_name: str):
+    def start_job(self, job_name: str, team_name: str, arguments: Dict):
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta.py:15]: must be Optional as arugments can be missing (none) 
Set default 
```suggestion
    def start_job(self, job_name: str, team_name: str, arguments: Optional[Dict] = None):
```

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/plugin_entry.py [hunk 6]
```diff
@@ -23,6 +23,7 @@ def run_job(context: JobContext) -> None:
         meta_job_runner.META_CONFIG = MetaPluginConfiguration(
             context.core_context.configuration
         )
+        meta_job_runner.ARGUMENTS = context.job_args.get_arguments()
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/plugin_entry.py:26]: What is this?

**Review by tozka — CHANGES_REQUESTED** [review]

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py [hunk 7]
```diff
@@ -49,12 +49,14 @@ def __init__(
         job_name: str,
         team_name: str,
         rest_api_url: str,
+        arguments: dict = None,
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py:52]: Should be optional 
```suggestion
        arguments: Optional[Dict] = None,
```

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py (hunk 2 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta_dag.py:29]: I got misled by the vdk flag --arguments thinking that if you pass an argument to a DAG job, it would propagate it to all the orchestrated jobs. Consequently, this was brought to the game so that when the runner invokes _vdk run <job> --arguments <args>_, the DAG would pass a default value to all the jobs if such is missing. If you think that such scenario won't he ever needed to be handled, I will remove this logic.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/plugin_entry.py (hunk 6 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/plugin_entry.py:26]: Related to https://github.com/vmware/versatile-data-kit/pull/1839#discussion_r1157123874
This way we get the args, passed by --arguments flag.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta.py (hunk 5 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/meta.py]: Done.

File: projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py (hunk 4 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py]: Added them but forgot an log assertion that the arguments are there. Will address this and push later.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py (hunk 3 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py]: Thanks! Changed those 4 to private.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py (hunk 1 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py]: Makes sense, my bad, changed it.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py (hunk 7 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job.py]: Done.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/plugin_entry.py (hunk 6 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/plugin_entry.py]: Removed it.

File: projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py (hunk 4 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py]: Added the assertions.

File: projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py [hunk 8]
```diff
@@ -327,6 +330,48 @@ def test_meta_job_duplicate_jobs(self):
...
+            self.runner = CliEntryBasedTestRunner(plugin_entry)
+            result = self._run_meta_job("meta-job-empty-arguments")
+            cli_assert_equal(0, result)
+            job2_post_req = [
+                req
+                for req, res in self.httpserver.log
+                if req.method == "POST"
+                and req.path.split("/jobs/")[1].split("/")[0] == "job2"
+            ]
+            job2_arguments = job2_post_req[0].json["args"]
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py:371]: extraxft as a helper function _get_job_arguments(httpserver) for example. 

The same code is used in two places.

**Review by tozka — APPROVED** [review]

File: projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py (hunk 8 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py]: Done.

**Comment by tozka** [comment]
I suppose it's in separate issue to update the README?

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job_executor.py [hunk 9]
```diff
@@ -10,9 +11,11 @@
 
 
 class RemoteDataJobExecutor(IDataJobExecutor):
-    def start_job(self, job_name: str, team_name: str):
+    def start_job(self, job_name: str, team_name: str, arguments: Dict = None):
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job_executor.py:14]: nit: this seems shady - I can recall there is IJobArguments interface, and there should be usages within vdk-core since all data jobs have arguments. Why does not this apply to dags..?

Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/remote_data_job_executor.py:14]: Well, I guess I missed it. Thanks! Applied it now.

File: projects/vdk-plugins/vdk-meta-jobs/README.md [hunk 10]
```diff
@@ -38,7 +39,8 @@ def run(job_input):
 When defining a job to be run following attributes are supported:
 * **job_name**: required, the name of the data job
 * **team_name:**: optional, the team of the data job. If omitted , it will use the meta job's team
-* **fail_meta_job_on_error**: optional, default is true. if true, the meta job will abort and fail if the orchestrated job fails, if false, meta job won't fail and continue.
+* **fail_meta_job_on_error**: optional, default is true. If true, the meta job will abort and fail if the orchestrated job fails, if false, meta job won't fail and continue.
+* **arguments**: optional, the arguments that are passed to the data job. The arguments could be used for parameter substitution in queries.
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/README.md:43]: `The arguments could be used for parameter substitution in queries.` 

That seems like it might be confusing here. In what queries in the DAG? In context of a dag. Just passing argument to underlying orchestrated data job seem enough

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py [hunk 11]
```diff
@@ -16,6 +18,7 @@ class SingleJob:
     job_name: str
     team_name: str = None
     fail_meta_job_on_error: bool = True
+    arguments: IJobArguments = None
```
Comment (tozka) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py:21]: We use primitive types here only. IJobArguments is interface for arguments to be read from Job Input. 

The arguments itself are of type "dict" , we even validate there are of type dict. While IJobArguments is another interface that is not compatible with dict.

File: projects/vdk-plugins/vdk-meta-jobs/README.md (hunk 10 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/README.md]: Now that I look at it, it might be confusing, yes, changing it.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py (hunk 11 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py]: Makes sense. Done.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py [hunk 12]
```diff
@@ -141,6 +149,27 @@ def _validate_fail_meta_job_on_error(self, job: Dict):
                 f" is {type(job['fail_meta_job_on_error'])}. Expected type is bool.",
             )
 
+    def _validate_arguments(self, job: Dict):
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:152]: nit: It is strange to have `validate_arguments` that accepts job in favour of the arguments to validate. It is also unexpected to have data job object of type dict. 

When designing method contracts, the presumption is to require for passing the minimum needed. Job object seems too broad of a scope to use, for simply validating its arguments. For example, why is the entire job object requested, while using one job field in particular?

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py [hunk 13]
```diff
@@ -141,6 +149,27 @@ def _validate_fail_meta_job_on_error(self, job: Dict):
...
+        if "arguments" in job:
+            if not isinstance(job["arguments"], dict):
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:154]: nit: were the arguments refactored to IJobArguments?

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py [hunk 14]
```diff
@@ -16,6 +16,7 @@ class SingleJob:
     job_name: str
     team_name: str = None
     fail_meta_job_on_error: bool = True
+    arguments: dict = None
     depends_on: List[str] = field(default_factory=list)
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py:20]: nit: What is the commonality of SingleJob and IStandaloneDataJob? It is strange that the DagJob isn't already relying on vdk-core API. What's the data job job top-level interface?
That question is out of scope for the current PR, yet I would like to revise it, in terms of refining the domain model.

**Review by ivakoleva — APPROVED** [review]

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py (hunk 14 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/api/meta_job.py:20]: Well, the commonality seems to be the job name and job arguments. SingleJob has the team name fail_meta_job_on_error and the depends_on list, meaning that SingleJob, IMO, could rely on IStandaloneDataJob. I will discuss this topic with the DAG-involved personas.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py (hunk 12 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:152]: The entire object is requested because when we raise the error, we log the job name. Another solution for this method would be to pass the args and job name. I can change it if this makes more sense.

File: projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py (hunk 13 above)
Comment (yonitoo) [reply projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/dag_validator.py:154]: Yes, but in the base class SingleJob the arguments are required to be of primitive type (dict in this case) and that's why we expect the arguments to be of type dict here.

**Review by ivakoleva — APPROVED** [review]
LGTM!

**Review by gageorgiev — APPROVED** [review]
looks good, I would just ask that you rename the test job from 'meta-job-arguments' to 'dag-arguments' and the step in it from 'meta.py' to 'dag.py', rest looks fine
